src/moonlink/src/storage/wal.rs

```diff
@@ -1 +1,331 @@
 pub(crate) mod wal_persistence_metadata;
+use crate::row::MoonlinkRow;
+use crate::storage::filesystem::accessor::base_filesystem_accessor::BaseFileSystemAccess;
+use crate::storage::filesystem::accessor::filesystem_accessor::FileSystemAccessor;
+use crate::storage::filesystem::filesystem_config::FileSystemConfig;
+use crate::table_notify::TableEvent;
+use crate::Result;
+use futures::stream::{self, Stream};
+use futures::{future, StreamExt};
+use serde::{Deserialize, Serialize};
+use std::pin::Pin;
+use std::sync::Arc;
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub(crate) enum WalEventEnum {
+    Append {
+        row: MoonlinkRow,
+        xact_id: Option<u32>,
+        is_copied: bool,
+    },
+    Delete {
+        row: MoonlinkRow,
+        xact_id: Option<u32>,
+    },
+    Commit {
+        xact_id: Option<u32>,
+    },
+    StreamAbort {
+        xact_id: u32,
+    },
+    StreamFlush {
+        xact_id: u32,
+    },
+}
+
+pub(crate) struct PersistAndTruncateResult {
+    pub file_persisted: Option<WalFileInfo>,
+    pub highest_deleted_file: Option<WalFileInfo>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct WalFileInfo {
+    pub file_number: u64,
+    pub highest_lsn: u64,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub(crate) struct WalEvent {
+    pub lsn: u64,
+    pub event: WalEventEnum,
+}
+
+impl WalEvent {
+    pub fn new(table_event: &TableEvent, last_highest_lsn: u64) -> Self {
+        // Try to get LSN from the table event, fallback to last_highest_lsn for events without LSN
+        let lsn = table_event
+            .get_lsn_for_ingest_event()
+            .unwrap_or(last_highest_lsn);
+
+        let event = match table_event {
+            TableEvent::Append {
+                row,
+                xact_id,
+                is_copied,
+                ..
+            } => WalEventEnum::Append {
+                row: row.clone(),
+                xact_id: *xact_id,
+                is_copied: *is_copied,
+            },
+            TableEvent::Delete { row, xact_id, .. } => WalEventEnum::Delete {
+                row: row.clone(),
+                xact_id: *xact_id,
+            },
+            TableEvent::Commit { xact_id, .. } => WalEventEnum::Commit { xact_id: *xact_id },
+            TableEvent::StreamAbort { xact_id } => WalEventEnum::StreamAbort { xact_id: *xact_id },
+            TableEvent::StreamFlush { xact_id } => WalEventEnum::StreamFlush { xact_id: *xact_id },
+            _ => {
+                unimplemented!("Invalid table event for WAL: {:?}", table_event)
+            }
+        };
+
+        Self { lsn, event }
+    }
+
+    pub fn into_table_event(self) -> TableEvent {
```

Comment:
Also curious about this, `WalEvent` is with visibility `pub(crate)` but `into_table_event` is exposed as `pub`, wondering why is that?